src/hash_to_field.cairo

```diff
@@ -0,0 +1,311 @@
+    //
+    // Parameters:
+    // - msg: a byte string containing the message to hash, chunked in big-endian 4-bytes
+    // - msg_bytes_len: length of the message in bytes
+    // - count: the number of field elements to output
+    //
+    // Returns:
+    // - fields: an array of field elements [u_0, ..., u_(count - 1)]
+    func hash_to_field{
+        range_check_ptr,
+        bitwise_ptr: BitwiseBuiltin*,
+        range_check96_ptr: felt*,
+        add_mod_ptr: ModBuiltin*,
+        mul_mod_ptr: ModBuiltin*,
+        sha256_ptr: felt*,
+        pow2_array: felt*,
+    }(msg: felt*, msg_bytes_len: felt, count: felt) -> (fields: UInt384**) {
+        alloc_locals;
+
+        let n_bytes = count * CURVE_M * CURVE_L;
+        let (result) = expand_msg_xmd(msg=msg, msg_bytes_len=msg_bytes_len, n_bytes=n_bytes);
+        let (result_fields: UInt384**) = alloc();
+        with result_fields {
+            hash_to_field_inner(expanded_msg=result, count=count, index=0);
+        }
+
+        return (fields=result_fields);
+    }
+
+    // Inner recursive function for hash_to_field
+    // Processes each count required in the hash_to_field function
+    func hash_to_field_inner{
+        range_check96_ptr: felt*,
+        add_mod_ptr: ModBuiltin*,
+        mul_mod_ptr: ModBuiltin*,
+        pow2_array: felt*,
+        result_fields: UInt384**,
+    }(expanded_msg: felt*, count: felt, index: felt) {
+        alloc_locals;
+
+        if (count == index) {
+            return ();
+        }
+
+        let offset = index * CURVE_L_IN_FELTS * CURVE_M;
+        let (fields: UInt384*) = alloc();
+        with fields {
+            hash_to_field_inner_inner(
+                expanded_msg=expanded_msg, count_index=index, degree_index=0, offset=offset
+            );
+        }
+        assert result_fields[index] = fields;
+
+        return hash_to_field_inner(expanded_msg=expanded_msg, count=count, index=index + 1);
+    }
+
+    // Innermost recursive function for hash_to_field
+    // Runs CURVE_M times (degree of the extension field)
+    func hash_to_field_inner_inner{
+        range_check96_ptr: felt*,
+        add_mod_ptr: ModBuiltin*,
+        mul_mod_ptr: ModBuiltin*,
+        pow2_array: felt*,
+        fields: UInt384*,
+    }(expanded_msg: felt*, count_index: felt, degree_index: felt, offset: felt) {
+        if (degree_index == CURVE_M) {
+            return ();
+        }
+
+        let (result) = _u512_mod_p(expanded_msg + offset);
+        // %{ print(hex(ids.result.d3 * 2**288 + ids.result.d2 * 2**192 + ids.result.d1 * 2**96 + ids.result.d0)) %}
+        assert fields[degree_index] = result;
+
+        return hash_to_field_inner_inner(
+            expanded_msg=expanded_msg,
+            count_index=count_index,
+            degree_index=degree_index + 1,
+            offset=offset + CURVE_L_IN_FELTS,
+        );
+    }
+
+    // Converts a 512-bit byte array to UInt384 and calls u512_mod_p
```

Comment:
Specify the expected form of the array :  
 - v15 v14 ... v1 v0 where u512 = v_i * (2^32)^i 
 - 0<=vi<2^32.